src/Runner.Worker/Handlers/CompositeActionHandler.cs

```diff
@@ -53,15 +78,40 @@ public async Task RunAsync(ActionRunStage stage)
                     childScopeName = $"__{Guid.NewGuid()}";
                 }
 
+
+                // If we need to setup containers beforehand, do it
+                List<JobExtensionRunner> containerDownloadSteps = null;
+                if (Data.ContainerSetupSteps != null && Data.ContainerSetupSteps.Count > 0)
+                {
+                    foreach (var step in Data.ContainerSetupSteps)
+                    {
+                        ArgUtil.NotNull(step, step.DisplayName);
+                        Guid stepId = Guid.NewGuid();
+                        step.ExecutionContext = ExecutionContext.CreateEmbeddedChild(childScopeName, Guid.NewGuid().ToString(), stepId);
+                    }
+                    containerDownloadSteps = Data.ContainerSetupSteps;
+                }
+
                 // Create embedded steps
                 var embeddedSteps = new List<IStep>();
-                foreach (Pipelines.ActionStep stepData in Data.Steps)
+
+                // Add container download steps
+                if (containerDownloadSteps != null && containerDownloadSteps.Count > 0)
+                {
+                    foreach(var step in containerDownloadSteps)
+                    {
+                        embeddedSteps.Add(step);
+                    }
+                }
+
+                foreach (Pipelines.ActionStep stepData in steps)
                 {
                     var step = HostContext.CreateService<IActionRunner>();
                     step.Action = stepData;
                     step.Stage = stage;
                     step.Condition = stepData.Condition;
-                    step.ExecutionContext = ExecutionContext.CreateEmbeddedChild(childScopeName, stepData.ContextName);
+                    ExecutionContext.Root.EmbeddedIntraActionState.TryGetValue(step.Action.Id, out var intraActionState);
+                    step.ExecutionContext = ExecutionContext.CreateEmbeddedChild(childScopeName, stepData.ContextName, step.Action.Id, intraActionState: intraActionState);
                     step.ExecutionContext.ExpressionValues["inputs"] = inputsData;
                     step.ExecutionContext.ExpressionValues["steps"] = ExecutionContext.Global.StepsContext.GetScope(childScopeName);
 
```

Comment:
Discussed offline...

might need to delete `ExecutionContext.Global.StepsContext.ClearScope(childScopeName);` further below due to post steps. Author may reference a prev step output, when specifying inputs for a step. So when we re-evaluate the inputs when the post step runs, outputs will be cleared already.

if we're not clearing outputs anymore, we might want to be cognizant of how much memory we are using. Consider modifying StepsContext to create an instance of `Sdk/DTExpressions2/Expressions2/Sdk/MemoryCounter.cs` and track every time we add/overwrite an output. If exceeds 100mb or something like that, trace to host log or debug trace to execution context.